### PR TITLE
feat: Add support for async queries

### PR DESCRIPTION
**Pull request by ca0904**
# feat: Add support for async queries

## Summary

This PR implements support for asynchronous query execution in fakesnow, enabling the use of `RESULT_SCAN()` function to retrieve results from previously executed queries using their SFQID (Snowflake Query ID).

**Addresses**: [Issue #236](https://github.com/tekumara/fakesnow/issues/236)

**Note**: This implementation solves the async query problem for `fakesnow.server()` usage but does not address the issue for `fakesnow.patch()` mode.

### Core Infrastructure
- **Results Caching**: Added a `results_cache` dictionary to store query results by SFQID in `FakeSnow` instance
- **Connection Updates**: Modified `FakeSnowflakeConnection` to accept and store results cache reference
- **Query Result Storage**: Implemented automatic caching of query results (arrow table, row count, SQL, parameters, transformed query) in cursor execution

### New Functionality
- **RESULT_SCAN Transform**: Added `result_scan()` transform in `transforms.py` to handle `SELECT * FROM TABLE(RESULT_SCAN('sfqid'))` queries
- **Result Retrieval**: Implemented logic in cursor to fetch cached results instead of executing new queries when `RESULT_SCAN` is used
- **Error Handling**: Added proper error handling for queries with missing SFQIDs (ProgrammingError with errno 2003)

### Server Support
- **Monitoring Endpoint**: Added `/monitoring/queries/{sfqid}` endpoint to check query status
- **Query Status**: Returns SUCCESS status for cached queries, 404 for missing queries

### Testing
- **Server Tests**: Added comprehensive tests for monitoring endpoint and error scenarios
- **Transform Tests**: Added tests for `RESULT_SCAN` transform functionality and edge cases

## Technical Details

The implementation follows Snowflake's async query pattern where:
1. Queries are executed and results are cached with a unique SFQID
2. Results can be retrieved later using `RESULT_SCAN('sfqid')`
3. Query status can be monitored via REST API
4. Proper error handling for missing or invalid SFQIDs

## Testing

All new functionality includes comprehensive test coverage:
- Unit tests for transform logic
- Integration tests for cursor behavior
- Server endpoint tests
- Error condition testing

## Breaking Changes

None. This is a new feature that extends existing functionality without modifying current behavior.

### CODE REVIEW

**Comment by ca0904** [comment]
Reson of failure -
```
FAILED tests/test_cursor.py::test_get_result_batches - TypeError: Can't instantiate abstract class FakeResultBatch with abstract method populate_data
FAILED tests/test_cursor.py::test_get_result_batches_dict - TypeError: Can't instantiate abstract class FakeResultBatch with abstract method populate_data
```

Should I make some changes for this?

File: fakesnow/cursor.py [hunk 1]
```diff
@@ -573,3 +574,6 @@ def to_pandas(self) -> pd.DataFrame:
 
     def to_arrow(self) -> pyarrow.Table:
         raise NotImplementedError()
+
+    def populate_data(self, connection: SnowflakeConnection | None = None, **kwargs: Any) -> ResultBatch:
```
Comment (tekumara) [inline fakesnow/cursor.py:578]: this was recently introduced in snowflake-connector-python 3.16.0 and now fixed on main

File: tests/test_server.py [hunk 2]
```diff
@@ -430,3 +439,101 @@ def test_server_types(scur: snowflake.connector.cursor.SnowflakeCursor) -> None:
             2,
         )
     ]
+
+
+def test_server_monitoring_query(scur: snowflake.connector.cursor.SnowflakeCursor, server: dict) -> None:
```
Comment (tekumara) [inline tests/test_server.py:444]: could we replace this with a test that uses `execute_async`? I think it captures end user behaviour better

File: fakesnow/server.py [hunk 3]
```diff
@@ -28,8 +29,11 @@
 logger.handlers = logging.getLogger("uvicorn").handlers
 logger.setLevel(logging.INFO)
 
+
 shared_fs = FakeSnow()
 sessions: dict[str, FakeSnowflakeConnection] = {}
+# Global mapping for query results by sfqid
+sfqid_results: dict[str, dict] = {}
```
Comment (tekumara) [inline fakesnow/server.py:36]: i think probably the best place for this is in `instance.FakeSnow`, stored as a `dict[str, pyarrow.Table]` .. this would allow it to be used by the patching in a future PR

**Review by tekumara — COMMENTED** [review]
Thank you @ca0904! I've made a couple of suggestions.

File: fakesnow/cursor.py (hunk 1 above)
Comment (ca0904) [reply fakesnow/cursor.py:578]: okay, thanks

File: tests/test_server.py (hunk 2 above)
Comment (ca0904) [reply tests/test_server.py:444]: okay, changing `cur.excuete()` to `cur.execute_async()`.

File: tests/test_server.py [hunk 4]
```diff
@@ -430,3 +430,42 @@ def test_server_types(scur: snowflake.connector.cursor.SnowflakeCursor) -> None:
...
+        # Test that we can also execute regular queries and retrieve their results
+        cur.execute("select 'regular query' as type")
+        regular_sfqid = cur.sfqid
+        assert regular_sfqid is not None
+        assert regular_sfqid != async_sfqid
+
+        # Test retrieving results from regular query sfqid
+        cur.get_results_from_sfqid(regular_sfqid)
+        regular_results = cur.fetchall()
+        assert regular_results == [("regular query",)]
```
Comment (tekumara) [inline tests/test_server.py:461]: this is already covered elsewhere, so we don't need it 

but we do need check queries returning no results, eg:
```python
    cur.execute_async("SELECT 1 as a, 2 as b where 1=0")
    assert cur.sfqid
    cur.get_results_from_sfqid(cur.sfqid)
    async_results = cur.fetchall()
    assert async_results == []
 ```
 
this won't work at the moment .. we probably need to store `cur._describe_last_sql()` in the dict and reuse that instead of calling `arrow_schema_to_describe_format`

File: fakesnow/conn.py [hunk 5]
```diff
@@ -29,6 +30,7 @@ def __init__(
         create_schema: bool = True,
         db_path: str | os.PathLike | None = None,
         nop_regexes: list[str] | None = None,
+        sfqid_results: dict[str, pyarrow.Table] | None = None,
```
Comment (tekumara) [inline fakesnow/conn.py:33]: should never be `None` - it needs to be created and owned outside so it can be shared across connections
```suggestion
results_cache: dict[str, pyarrow.Table],
```
also `results_cache` is what snowflake calls this

File: fakesnow/server.py [hunk 6]
```diff
@@ -96,6 +123,51 @@ async def query_request(request: Request) -> JSONResponse:
 
         expr = parse_one(sql_text, read="snowflake")
 
+        result_scan_match = re.match(
```
Comment (tekumara) [inline fakesnow/server.py:126]: could this be handled inside `cursor._execute`? then we wouldn't need the regex and the special path here

File: fakesnow/server.py [hunk 7]
```diff
@@ -198,6 +270,49 @@ async def session(request: Request) -> JSONResponse:
...
+        sf_table = to_sf(arrow_table, rowtype)
+        batch_bytes = to_ipc(sf_table)
+        rowset_b64 = b64encode(batch_bytes).decode("utf-8")
+
+        return JSONResponse(
+            {
+                "data": {
+                    "parameters": [
+                        {"name": "TIMEZONE", "value": "Etc/UTC"},
+                    ],
```
Comment (tekumara) [inline fakesnow/server.py:295]: not needed
```suggestion
```

File: fakesnow/server.py [hunk 8]
```diff
@@ -198,6 +270,49 @@ async def session(request: Request) -> JSONResponse:
...
+                    "queries": [{"status": "SUCCESS", "sfqid": sfqid}],
+                    "rowtype": rowtype,
+                    "rowsetBase64": rowset_b64,
+                    "total": arrow_table.num_rows,
+                    "queryId": sfqid,
+                    "queryResultFormat": "arrow",
+                    "finalDatabaseName": conn.database,
+                    "finalSchemaName": conn.schema,
```
Comment (tekumara) [inline fakesnow/server.py:303]: not needed - this just returns status, not results
```suggestion
```

File: fakesnow/server.py [hunk 9]
```diff
@@ -198,6 +270,49 @@ async def session(request: Request) -> JSONResponse:
...
+        batch_bytes = to_ipc(sf_table)
+        rowset_b64 = b64encode(batch_bytes).decode("utf-8")
+
+        return JSONResponse(
+            {
+                "data": {
+                    "parameters": [
+                        {"name": "TIMEZONE", "value": "Etc/UTC"},
+                    ],
+                    "queries": [{"status": "SUCCESS", "sfqid": sfqid}],
```
Comment (tekumara) [inline fakesnow/server.py:296]:
```suggestion
"queries": [{"status": "SUCCESS"}],
```

File: fakesnow/server.py [hunk 10]
```diff
@@ -198,6 +270,49 @@ async def session(request: Request) -> JSONResponse:
...
+        if not (arrow_table := conn.sfqid_results.get(sfqid)):
+            raise ServerError(status_code=404, code="390105", message=f"Query with SFQID {sfqid} not found.")
+
+        describe_results = arrow_schema_to_describe_format(arrow_table.schema)
+        rowtype = describe_as_rowtype(describe_results)
+
+        # Convert to Snowflake format and encode
+        sf_table = to_sf(arrow_table, rowtype)
+        batch_bytes = to_ipc(sf_table)
+        rowset_b64 = b64encode(batch_bytes).decode("utf-8")
```
Comment (tekumara) [inline fakesnow/server.py:288]:
```suggestion
```

**Review by tekumara — COMMENTED** [review]
making progress!

File: tests/test_server.py (hunk 4 above)
Comment (ca0904) [reply tests/test_server.py:461]: this assert was added to pass `pyright` as `sfqid` is `str | None`.

Comment (ca0904) [reply tests/test_server.py:461]: should i make `results_cache: dict[str, Tuple[pyarrow.Table, list]]` for storing `cur._describe_last_sql()` as well.

File: fakesnow/server.py (hunk 6 above)
Comment (ca0904) [reply fakesnow/server.py:126]: should i add function in transform for this?

**Review by ca0904 — COMMENTED** [review]
Hi @tekumara! Please take a look.

**Review by tekumara — COMMENTED** [review]
looks fantastic! ... i just need to verify the error cases on a real snowflake instance

**Comment by ca0904** [comment]
Hey! I was wondering when this functionality will be released.

**Comment by tekumara** [comment]
@ca0904 I've pushed some tidy up and aligned the error handling with snowflake, and will now make a release.

**Comment by tekumara** [comment]
@ca0904 I've pushed some tidy up and aligned the error handling with snowflake, and will now make a release. Thank you for your great contribution!
